### PR TITLE
Implemented project slugs

### PR DESCRIPTION
**Pull request by abstractcoder**
First pass at project slugs

I have very little experience using database triggers but it felt right in this case.

I'm having a hard time figuring out how to support the old project url format (with ids) and the new project urls (with slugs) at the same time. I'm also having a hard time figuring out how to set the project show url to use slugs (since thats the most important one), without a huge amount of effort. I wish Phoenix used the symbol in the url as the key (i.e. :id or :slug) when using the path helpers. As it is you can overwrite which field is used when you pass a struct to a path helper or you can specifically call project.slug when using the path helper. There is commented out code that does a redirect from show with id to show with slug but that broke a bunch of tests and feels wrong.

I had to use naivedate_time_usec for the project_slugs timestamps to get the tests to work since the database trigger deletes any slugs beyond the 10 most recently used and I didn't want to slow down the tests by adding sleeps. This feels bad but I can't think of a better solution.

Deleting the oldest project_slugs also doesn't quite make sense, because we really only need to save the slugs that appeared in urls the project owner has shared, but thats a whole other issue.

On a positive note, I'm really happy with how I was able to incorporate the data migration into the database migration, so that the slug column is setup to not allow nulls.

### CODE REVIEW

**Comment by abstractcoder** [comment]
I pushed an update that sets Project to use the slug as the default param. I then created a route for the old project id path only for the show action. This feels like two steps forward and one step back but its progress :)

**Comment by nickgartmann** [comment]
The DB trigger here is a really clever solution, I've not really used them so can't speak to their goodness or badness, but I agree that doing this as a check in elixir would be _weird_. 

I think not redirecting off of permalinked projects is fine, since we aren't trying to solve digital transmission of projects with slugs, we're trying to make them easy to type into a URL bar if you heard about them in person (or are typing from a link displayed on a screen). 

I'm maybe not understanding why the system you have with the different URLs for params is problematic, might be we just need to talk it out on our call on friday?

**Comment by abstractcoder** [comment]
I'm both trying to solve for manually typing a url, but also improving the experience of sharing a link (Social Cards may have a bigger impact on this).

I made a few updates since my original comment and I feel better about the current routing situation. Now slugs are the default, but I included a fallback id route to continue supporting links that have already been shared.

File: lib/digital_public_works/projects.ex [hunk 1]
```diff
@@ -103,11 +106,23 @@ defmodule DigitalPublicWorks.Projects do
       ** (Ecto.NoResultsError)
 
   """
-  def get_project!(id), do: Repo.get!(Project, id) |> Repo.preload([:user, :organizations])
+  def get_project!(id), do: Repo.get!(Project, id) |> preload()
```
Comment (nickgartmann) [inline lib/digital_public_works/projects.ex:109]: I THINK this works

`Repo.get!(preload(Project), id)`

**Review by nickgartmann — APPROVED** [review]

**Comment by abstractcoder** [comment]
I added the comment to project_slug.ex.

I looked at moving get_projects_by_url! to the controller but didn't want to have to rewrite the unit tests as controller tests so I left it as is for now.

I looked at changing how preload works to reduce the number of queries but didn't see a way to get it working with both Repo.get! and Repo.all.

I'm going to push it live.

**Pull request merged**